morango/syncsession.py

```diff
@@ -220,56 +270,154 @@ class SyncClient(object):
+            try:
+                response = self.sync_connection._create_transfer_session(data)
+            except requests.HTTPError as e:
+                self.current_transfer_session.delete(soft=True)
+                raise
+
+            self.current_transfer_session.server_fsic = response.json().get('server_fsic') or '{}'
+            self.current_transfer_session.records_total = response.json().get('records_total')
+            self.current_transfer_session.transfer_stage = transfer_status.PULLING
+            self.current_transfer_session.save()
 
-    def initiate_push(self, sync_filter):
+    def _push_records(self, callback=None):
+        buffered_records = Buffer.objects.filter(transfer_session=self.current_transfer_session).order_by('pk')
 
-        self._create_transfer_session(True, sync_filter)
+        while self.current_transfer_session.records_transferred < self.current_transfer_session.records_total:
+            # slice queryset to get chunk of records for sending
+            records = buffered_records[self.current_transfer_session.records_transferred:self.chunk_size + self.current_transfer_session.records_transferred]
+            # serialize and send records to server
+            serialized_recs = BufferSerializer(records, many=True)
+            logger.info('Pushing {} records at a time with {}/{} records transferred'.format(self.chunk_size,
+                                                                                             self.current_transfer_session.records_transferred,
+                                                                                             self.current_transfer_session.records_total))
+            try:
+                self.sync_connection._push_record_chunk(serialized_recs.data)
+            except requests.HTTPError as e:
+                self._close_transfer_session()
+                raise
 
-        _queue_into_buffer(self.current_transfer_session)
+            # update records_transferred upon successful request
+            self.current_transfer_session.records_transferred += self.chunk_size
+            # ensure records transferred does not go over total records
+            if self.current_transfer_session.records_transferred > self.current_transfer_session.records_total:
+                self.current_transfer_session.records_transferred = self.current_transfer_session.records_total
+            self.current_transfer_session.save()
 
-        # update the records_total for client and server transfer session
-        records_total = Buffer.objects.filter(transfer_session=self.current_transfer_session).count()
-        if records_total == 0:
+    def _pushing(self):
+        try:
+            self.sync_connection._update_transfer_session({'records_total': self.current_transfer_session.records_total},
+                                                          self.current_transfer_session)
+        except requests.HTTPError as e:
             self._close_transfer_session()
```

Comment:
What types of errors would trigger this? And does closing the transfer session allow us to resume it after?

Reply:
HTTPErrors are usually 4xx errors or 5xx errors. So my assumption is that something is bad with the transfer session request so we just close it because it probably won't work the second time. And no we can not resume it because we set active to False.